### ISSUE TITLE
Add support for multiple independent entities namespaces

### ISSUE THREAD

**Issue by Dishwasha** [vuex-orm/vuex-orm#360]
A larger single page application may be composed of a number of features that may be better served by having isolation between the models and overlapping entity types to reduce unintended side effects.  Currently VuexORM supports a single database and Vuex entities namespace which is configured globally.  Extending this to support registering multiple VuexORM databases and linking models to those namespaces, or allowing a single database to have multiple namespaces that models can be linked to would be two possible ways to solve this.

**Comment by kiaking** [comment]
@Dishwasha Thank you for the feedback! Yes, I think that make sense. I'll mark this as an enhancement and see if we can introduce such feature.

**Comment by Dishwasha** [comment]
I would like to assist in the development of this feature.  Do you have any design suggestions that I could begin researching?

**Comment by kiaking** [comment]
@Dishwasha Thank you so much! OK so the internal architecture is like this right now.

The database instance that holds all models, are stored at singleton object called Container. Now, when Vuex ORM needs to get Model class, it accesses the Container and get database. Something like this.

```js
import Container from './Container'

function getModel (name) {
  return Container.database.getModel(name)
}
```

Mainly, Query class needs to have access to Container, because Query class must resolve Model by string. This is because Query class is used through Vuex Module such as actions and getters. Fr example, Vuex Getters will call Query class like this.

```js
query: (state: RootState) => (entity: string): Query => {
  return new Query(state, entity)
}
```

See that `entity` is a string, and Query must fetch actual Model class from that string.

So the hard part is, Vuex ORM must know in this case from which database should Query class resolve the Model. And how can we tell if the model is in different namespace. We need to have some smart design to tackle this part.

**Comment by Dishwasha** [comment]
Database seems fairly well isolated to an entity namespace.  Do you think it makes sense to change Container to support registering multiple databases mapped by entity namespace (perhaps making a sensible default be the first one or the one mapped to the default 'entities' namespace) and then methods like Query.database knows how to look up the correct database by namespace?   Or would having multiple Containers be better?

**Comment by kiaking** [comment]
> Do you think it makes sense to change Container to support registering multiple databases mapped by entity namespace

Yes, I think that's the first thing we should do!

> and then methods like Query.database knows how to look up the correct database by namespace?

Yes I think this is true too. We must also look into Model as well because it has access to database as well. Query class could be relatively easier because it always gets the state object through Vuex Modules. But Model's don't. So maybe we need to do something with model when registering them to the database.

> Or would having multiple Containers be better?

I think a single Container is better. I mean it has to be global singleton object so it might not be possible to have multiple containers.

**Comment by georgechaduneli** [comment]
Any updates on this?

**Comment by kiaking** [comment]
Sorry, not at the moment. This is kinda big change internally, and need to do design before we can start... Haven't had time to tackle this one yet.

**Comment by kiaking** [comment]
I've created a new issue about more detailed design on implementing this issue at https://github.com/vuex-orm/vuex-orm/issues/515. Let's continue discussing over there! I would love to have feedback especially on the limitations section, whether that limitation is acceptable or not.
